README.md
### Welcome to the FIFA 22 Live Editor!

For now this tool allows you to:
- Mod almost **all** legacy files (add or replace)
- Edit Players
- Edit Teams
- Transfer & Loan Players
- Generate Minifaces
- Edit Database (only tables that are part of the squadfile)
- Execute custom lua scripts. Check [LUA API](https://github.com/xAranaktu/FIFA-22-Live-Editor/wiki/LUA-API)

My plans for Live Editor is to make all in one tool. 
Modding, editing game memory, editing database, API for developers and many other features that you never dreamed about.

Check [WIKI](https://github.com/xAranaktu/FIFA-22-Live-Editor/wiki) for Guides and Tutorials

***

If you want to download the tool then DONT CLONE THE REPO...
Get it from releases:

![](https://i.imgur.com/ocPeJiV.png)

***

![](https://i.imgur.com/SSCMdKs.png)
![](https://i.imgur.com/FCbP4mC.png)
![](https://i.imgur.com/JFljI3A.png)
![](https://i.imgur.com/pgQStaD.png)
![](https://i.imgur.com/mH7G1gO.png)
![](https://i.imgur.com/etIRoCG.png)


changelog.txt
v22.1.1.9:
    - Reworked DirectX hooks. (Hopefully DX12 will be more stable now)
    - Fixed:
        - Some File Browser crashes
        - Crashes in Teams Editor if teamkittypedtechid > 3
        - Saving alpha channel color in Teams Editor -> teamkits

v22.1.1.8:
    - Added names for a lot of IDs in Players Editor (Thanks Chlipi667!)
    - Fixed:
        - editing player goals scored, assists & clean sheets
        - names of the competitions displayed while editing player stats
        - jerseysleevelengthcode values in players editor
        - jerseystylecode for ID 2 ("Dark Brown" -> "Half Tuck")

v22.1.1.7:
    - Added match-fixing for single fixtures.
    - Added Home & Away standings.
    - Added option to edit goals for, goals against, wins, draws and losses in standings.
    - Fixed problem with reverting attributes for some players in manager career mode.

v22.1.1.6:
    - Fixed game crashes that were occurring when loading preseason tournaments standings.

v22.1.1.5:
    - Added to "Teams Editor":
        - "Standings"
        - "Fixtures"
    - Fixed loading "notfound" image for minifaces, flags & crests.
    - Fixed creating transfer/loan offer in player career mode. 

v22.1.1.4:
    - Adding checking transfer ban file permissions. (The game won't crash now, but bans won't be saved to file until you fix the problem with permissions.)
    - Fixed loading transfer bans when loading career saves multiple times.

v22.1.1.3:
    - Added Transfer Ban System.
    - Added LUA functions:
        - AddTransferBan
        - RemoveTransferBan
        - IsTeamTransferBanned

v22.1.1.2:
    - Added "File Browser" (Legacy files only atm.)

v22.1.1.1:
    - Fixed detecting player injury if player is seriously injured (till now the tool was only able to recognize light injures)
    - Fixed crashes when the dcplayernames table got no records
    - Fixed crashes on problem with reading player form (?)

v22.1.1.0:
    - "Transfer Player", "Loan Player" & "Terminate Loan" functions are now available also outside of Career Mode.
    - Added button allowing you to copy player name and/or id to clipboard in Players Editor
    - Added button allowing you to copy team name and/or id to clipboard in Teams Editor
    - Added LUA functions:
        - "DeletePlayer"
        - "CreatePlayer"
        - "PlayerExists"

v22.1.0.9:
    - Added LUA Scripts:
        - "export_season_stats.lua"
        - "extend_user_team_players_contracts.lua"
    - Added LUA functions:
        - "GetSaveUID"
        - "GetCompetitionNameByID"
        - "GetCompetitionNameByObjID"
        - "GetPlayersStats"
        - "GetPlayerStats"
    - Fixed crashes at startup "caused" by Medal recorder
    - Fixed crashes that occur when the first param passed to "Log" lua function was not a string

v22.1.0.8:
    - Added "Stats" to Players Editor (Goals, Assists & Clean Sheets)
    - Fixed width of the actions popup window in Database Editor
    - Removed unnecessary lua functions hooks

v22.1.0.7:
    - Importing miniface feature now works in fullscreen mode (replaced nativefiledialog with ImGuiFileDialog)
    - Root live editor mods folder can now be changed in settings.
    - Fixed crashes on importing/generating miniface when the target destination didn't exists.

v22.1.0.6:
    - Fixed creating transfer/loan offer for player in player career mode.

v22.1.0.5:
    - Added manager bodytypecodes 
    - Added "Kits" to Teams Editor

v22.1.0.4:
    - Added "Manager" to Teams Editor
    - Editing user team transferbudget in Teams Editor will now also update club finances
    - Fixed "Team Not Found" that occured for "Rest of World" league teams.
    - Minifaces can be now generated also outside of Career Mode.
    - Changed default camera when generating miniface from "Head and Shoulders" to "Head" (the same that were used in FIFA 21)

v22.1.0.3:
    - Added FIFA 22 Bodytypecodes
    - Fixed filtering rows (Fixes terminate loan bug and probably a few others)
    - Fixed creating transfer & loan offer for player career mode player

v22.1.0.2:
    - Added "Generate Miniface" to "Players Editor"
    - Added "Loan Player" to "Players Editor"

v22.1.0.1:
    - Added "Players Editor"
    - Added "Teams Editor"
    - Added "Database Editor"

v22.1.0.0:
    - Legacy files modding possibility.


lua/scripts/export_season_stats.lua
-- This script will export current season stats (goals scored, assists etc.) to CSV.
-- The SEASON_STATS.csv file will be created on Desktop.
-- The game resets stats at the beginning of every season, so you need to run the script at the end of every season.

local desktop_path = string.format("%s\\Desktop", os.getenv('USERPROFILE'))
local current_date = GetCurrentDate()
local file_path = string.format("%s\\SEASON_STATS_%d_%d_%d.csv", desktop_path, current_date.day, current_date.month, current_date.year)

-- Don't Touch
local columns = {
    "position",
    "playerid",
    "playername",
    --"team",
    "competition",
    "appearances",
    "AVG",
    "MOTMs",
    "goals",
    "assists",
    "yellow_cards",
    "two_yellow",
    "red_cards",
    "saves",
    "goals_conceded",
    "cleansheets",
}

local col_to_key = {
    position = "position",
    playerid = "playerid",
    playername = "playername",
    --team = "teamname",
    competition = "compname",
    appearances = "app",
    AVG = "avg",
    MOTMs = "motm",
    goals = "goals",
    assists = "assists",
    yellow_cards = "yellow",
    two_yellow = "two_yellow",
    red_cards = "red",
    saves = "saves",
    goals_conceded = "goals_conceded",
    cleansheets = "clean_sheets",
}

function get_pos_name(pos_id)
    POS_TO_NAME = {
        [0] = "GK",
        [1] = "SW",
        [2] = "RWB",
        [3] = "RB",
        [4] = "RCB",
        [5] = "CB",
        [6] = "LCB",
        [7] = "LB",
        [8] = "LWB",
        [9] = "RDM",
        [10] = "CDM",
        [11] = "LDM",
        [12] = "RM",
        [13] = "RCM",
        [14] = "CM",
        [15] = "LCM",
        [16] = "LM",
        [17] = "RAM",
        [18] = "CAM",
        [19] = "LAM",
        [20] = "RF",
        [21] = "CF",
        [22] = "LF",
        [23] = "RW",
        [24] = "RS",
        [25] = "ST",
        [26] = "LS",
        [27] = "LW"
    }
    return POS_TO_NAME[pos_id] or "INVALID"
end

function get_players()
    local result = {}
    
    local rows = GetDBTableRows("players")
    for i=1, #rows do
        local row = rows[i]
        local playerid = math.floor(row.playerid.value)
        
        local player = {}
        
        player.position = get_pos_name(math.floor(row.preferredposition1.value))
        
        -- Get the names later, only if needed as it's very slow
        player.name = ""
        --player.teamname = ""

        result[playerid] = player
    end
    
    return result
end

function get_stats(players)
    local result = {}
   
    local all_stats = GetPlayersStats()

    for i=1, #all_stats do
        local stat = all_stats[i]
        local playerid = stat.playerid
        local app = stat.app

        if playerid > 0 and playerid < 4294967295 and app > 0 then
            local player = players[playerid]
            if player ~= nil then
                if player.name == "" then
                    player.name = GetPlayerName(playerid)
                end
                
                local stats = player.stats or {}
                local compobjid = stat.compobjid
                stats[compobjid] = {}
                
                stats[compobjid]["playerid"] = playerid
                stats[compobjid]["playername"] = player.name
                stats[compobjid]["position"] = player.position
                
                stats[compobjid]["goals"] = stat.goals
                stats[compobjid]["yellow"] = stat.yellow
                stats[compobjid]["red"] = stat.red
                stats[compobjid]["assists"] = stat.assists
                stats[compobjid]["clean_sheets"] = stat.clean_sheets
                stats[compobjid]["compobjid"] = compobjid
                stats[compobjid]["compname"] = GetCompetitionNameByObjID(compobjid)
                
                -- Not sure
                stats[compobjid]["motm"] = stat.motm
                stats[compobjid]["saves"] = stat.saves
                stats[compobjid]["goals_conceded"] = stat.goals_conceded
                stats[compobjid]["two_yellow"] = stat.two_yellow
                
                local avg = stat.avg
                if app > 1 then 
                    avg = (avg / app) / 10 
                elseif app == 1 then
                    avg = avg / 10
                end
                
                stats[compobjid]["app"] = app
                stats[compobjid]["avg"] = string.format("%0.2f", avg)

                player.stats = stats
                result[playerid] = player

            end
        end
    end

    return result
end

assert(IsInCM(), "Script must be executed in career mode")
ReloadDB()

local file = io.open(file_path, "w+")
io.output(file)

io.write(table.concat(columns, ","))
io.write("\n")
local players_stats = get_stats(get_players())
for playerid, data in pairs(players_stats) do
    for compobjid, stat in pairs(data.stats) do
        local row = {}
        for i=1, #columns do
            local colname = columns[i]
            local _key = col_to_key[colname]
            table.insert(row, stat[_key])
        end
        io.write(table.concat(row, ","))
        io.write("\n")
    end

end

io.close(file)
Log("Done")
